tracee-ebpf/tracee/proc.go

```diff
@@ -0,0 +1,87 @@
+package tracee
+
+import (
+	"github.com/prometheus/procfs"
+)
+
+func gatherEntireProcessTree() (map[uint32][]uint32, error) {
+	fs, err := procfs.NewFS("/proc")
+	if err != nil {
+		return nil, err
+	}
+
+	procs, err := fs.AllProcs()
+	if err != nil {
+		return nil, err
+	}
+
+	processMap := map[uint32][]uint32{}
+	for i := range procs {
+		stat, err := procs[i].Stat()
+		if err != nil {
+			return nil, err
+		}
+		processMap[uint32(stat.PID)] = processMap[uint32(stat.PID)]
+		processMap[uint32(stat.PPID)] = append(processMap[uint32(stat.PPID)], uint32(stat.PID))
+	}
+
+	return processMap, nil
+}
+
+// populateProcessTreeFilterMap takes a map of the process tree (k=ppid, v=[]childpid)
+// and a filter specification map (k=pid,v=trace/not) and returns a fully populated map
+// where k=pid, v=trace/not for all pids
+func populateProcessTreeFilterMap(processTree map[uint32][]uint32,
```

Comment:
I don't understand why we shuold first gather the process tree representation, and only then initialize the map.
There is a big window for race conditions using this method, and it wastes cycles for processes we don't really need (were not requested by the user tree filter)

Reply:
The reason we gather the process tree first is so we can determine all of the descendant processes. Is there a way to query proc on a per pid basis to traverse through all its child pids that i'm not aware of?

Reply:
As in for each PID we can gather its PPID, but I don't see anything like a list of child pids

Reply:
Yes, there is actually a way to do that. To every process, you can list all the threads and their children using:
/proc/pid/task/tid/children

Reply:
Ahhh, thanks for teaching me that. Will have iteration #5 soon :-)

Reply:
@yanivagman What are your thoughts on the proc man page section on this file:

```
 /proc/[pid]/task/[tid]/children (since Linux 3.5)
        A  space-separated list of child tasks of this task.  Each child task is rep‐
        resented by its TID.

        This option is intended for use by the checkpoint-restore (CRIU) system,  and
        reliably  provides  a list of children only if all of the child processes are
        stopped or frozen.  It does not work properly if children of the target  task
        exit  while  the  file is being read!  Exiting children may cause non-exiting
        children to be omitted from the list.  This makes this  interface  even  more
        unreliable  than  classic  PID-based approaches if the inspected task and its
        children aren't frozen, and most code should probably not use this interface.
```

Also we'll still have to convert the TID to TGID, regardless of the stability concerns listed above.

Reply:
The current implementation (map entire proc tree to determine parent/child relationships) can have a data race where we record procs that exit after being set up (false positive). However since the map is built after events are installed the update in sched_process_exit should catch this (must confirm if this is reflected in the map).

If we use `/proc/[pid]/task/[tid]/children` we have the opposite problem (false negative) where by nature of procfs where we may not know about procs that do in fact exist. The map updates in sched_process_(fork|exit) won't handle this.

Reply:
> The current implementation (map entire proc tree to determine parent/child relationships) can have a data race where we record procs that exit after being set up (false positive). However since the map is built after events are installed the update in sched_process_exit should catch this (must confirm if this is reflected in the map).

The race is between the time you read procfs data, to the time you update the map. If you update the map and after some process already exited, you will never receieve sched_process_exit again for this process, which means that the pid will stay in the map. This is why I suggested to traverse procfs twice

Reply:
> @yanivagman What are your thoughts on the proc man page section on this file:

I see. Let's not use this interface then

Reply:
Ok, so here is another suggestion:
For every process 'p' in procfs:
1. Get its ppid
2. If ppid = one of the given tree filters pids -> add 'p' to the map, and continue to the next process
3. If ppid = 1 -> continue to the next process (as we reached the root of the process tree)
4. Go to step 1 with ppid 

This method will find the relevant set of pids which are part of the tree of the pid given by the tree filter, and update the map in (almost) the same time, reducing the window of a possible race condition

Reply:
@rafaeldtinoco to get the changes of the process tree in runtime is not the problem here, as we have sched_process_fork and sched_process_exit.
The problem here is - how can we initialize a map that contains a "snapshot" of the pids in the process tree whose root is some given pid. The challenge here is to initialize this map and avoid race conditions of exited/new processes in this process tree

Reply:
> @rafaeldtinoco to get the changes of the process tree in runtime is not the problem here, as we have sched_process_fork and sched_process_exit.

I had that in mind when I wrote.

> The problem here is - how can we initialize a map that contains a "snapshot" of the pids in the process tree whose root is some given pid. The challenge here is to initialize this map and avoid race conditions of exited/new processes in this process tree

That's why I referred a userland way of following the process tree (by following forks, execs and exits from userland, using netlink, in parallel to BPF program startup).

A taskstats/conector go routine could monitor selected pid trees for forks an and exits and keep updating the map by removing the pid if that exited before the BPF program was loaded (or update the map with other needs)

It would still be racy - as there is no mutex locking mechanism in between kernel BPF program and a userland task - but the fact that pid numbers would keep growing, instead of reusing just exited ones, would protect us until BPF program is loaded, as the race window is small (not bigger than pid # reusable).

Of course this ads complexity to the initial logic.

Another possibility would be to timestamp each entry in the map and have some kind of prune mechanism for leftovers.

Reply:
@yanivagman A couple of unfortunate issues with [your design](https://github.com/aquasecurity/tracee/pull/927#discussion_r699629208) that I've run into. If we use `filepath.WalkDir()` we're not able to jump to different points in the process tree, it's iterative. That can be worked around, but the reason it's iterative is because it reads the entire directory upfront, taking a snapshot of it, and presenting the same data race as the current design. Even if not using `filepath.WalkDir` I run into the same issue with similar implementations that aren't iterative, we can only make the race window smaller if we can't get atomic state of the process tree (i.e. halt the OS :smile: ). 

@rafaeldtinoco I will read through that kernel doc you linked, i'm not familiar with that interface. 

Another simple idea (again, still has unavoidable race) is to use the current design except re-check the state of the whole process tree on every iteration of traversing PID/PPID relationship (i.e. call `gatherEntireProcessTree()`)  in every iteration of the loop in `populateProcessTreeFilterMap()`. I may need to explore this.

Reply:
@grantseltzer its likely we wont use this, as a curiosity though: https://bewareofgeek.livejournal.com/2945.html ... I used something like this in my previous life example =). Its a good example on how to get PROC_EVENT_FORK, PROC_EVENT_EXIT, PROC_EVENT_EXEC from userland (before we could do this with eBPF, for example). It has a netlink socket buffer and its a stream from kernel to userland, but.. yet, its fast. 

> But again, I think your discussion with Yaniv is in advanced state already.